go/vt/vttablet/tabletserver/messager/message_manager.go

```diff
@@ -470,50 +508,28 @@ func (mm *messageManager) send(receiver *receiverWithStatus, qr *sqltypes.Result
 
 	// This is the cleanup.
 	defer func() {
-		// Discard messages from cache only at the end. This will
-		// prevent them from being requeued while they're being postponed.
-		mm.cache.Discard(ids)
-
-		mm.mu.Lock()
-		defer mm.mu.Unlock()
+		func() {
+			mm.mu.Lock()
+			defer mm.mu.Unlock()
 
-		receiver.busy = false
-		// Rescan if there were no previously available receivers
-		// because the current receiver became non-busy.
-		if mm.curReceiver == -1 {
-			mm.rescanReceivers(-1)
-		}
+			receiver.busy = false
+			// Rescan if there were no previously available receivers
+			// because the current receiver became non-busy.
+			if mm.curReceiver == -1 {
+				mm.rescanReceivers(-1)
+			}
+		}()
+
+		func() {
+			mm.streamMu.Lock()
+			defer mm.streamMu.Unlock()
+			// Discard messages from cache only at the end. This will
+			// prevent them from being requeued while they're being postponed.
+			mm.cache.Discard(ids)
+		}()
```

Comment:
It seems like you are nesting an extra level of functions just to be able to use defer. There aren't any exit paths inside either func, so it seems like you could eliminate the wrapping functions and defers by just moving both unlocks under the statement. I guess maybe that's safer in case of panics?

If you feel like the current code should remain, it's definitely worth a comment about the reasoning for the choice.